library/core/tests/fmt/num.rs

```diff
@@ -128,30 +128,375 @@ fn test_format_int_exp_precision() {
+    assert_eq!(format!("{:.2e}", -1085), "-1.09e3");
+    assert_eq!(format!("{:.2e}", -1065), "-1.07e3");
+    assert_eq!(format!("{:.2e}", -1045), "-1.05e3");
+    assert_eq!(format!("{:.2e}", -1025), "-1.03e3");
+    assert_eq!(format!("{:.2e}", -1005), "-1.01e3");
+    assert_eq!(format!("{:.2e}",  1005), "1.01e3" );
+    assert_eq!(format!("{:.2e}",  1025), "1.03e3" );
+    assert_eq!(format!("{:.2e}",  1045), "1.05e3" );
+    assert_eq!(format!("{:.2e}",  1065), "1.07e3" );
+    assert_eq!(format!("{:.2e}",  1085), "1.09e3" );
+    assert_eq!(format!("{:.2e}",  1105), "1.11e3" );
+    assert_eq!(format!("{:.2e}",  1125), "1.13e3" );
+    assert_eq!(format!("{:.2e}",  1145), "1.15e3" );
+    assert_eq!(format!("{:.2e}",  1165), "1.17e3" );
+    assert_eq!(format!("{:.2e}",  1185), "1.19e3" );
+    assert_eq!(format!("{:.2e}",  1205), "1.21e3" );
+    assert_eq!(format!("{:.2e}",  1225), "1.23e3" );
+    assert_eq!(format!("{:.2e}",  1245), "1.25e3" );
+    assert_eq!(format!("{:.2e}",  1265), "1.27e3" );
+    assert_eq!(format!("{:.2e}",  1285), "1.29e3" );
+    assert_eq!(format!("{:.2e}",  1305), "1.31e3" );
+    assert_eq!(format!("{:.2e}",  1325), "1.33e3" );
+    assert_eq!(format!("{:.2e}",  1345), "1.35e3" );
+    assert_eq!(format!("{:.2e}",  1365), "1.37e3" );
+    assert_eq!(format!("{:.2e}",  1385), "1.39e3" );
+    assert_eq!(format!("{:.2e}",  1405), "1.41e3" );
+    assert_eq!(format!("{:.2e}",  1425), "1.43e3" );
+    assert_eq!(format!("{:.2e}",  1445), "1.45e3" );
+    assert_eq!(format!("{:.2e}",  1465), "1.47e3" );
+    assert_eq!(format!("{:.2e}",  1485), "1.49e3" );
+    assert_eq!(format!("{:.2e}",  1505), "1.51e3" );
+    assert_eq!(format!("{:.2e}",  1525), "1.53e3" );
+    assert_eq!(format!("{:.2e}",  1545), "1.55e3" );
+    assert_eq!(format!("{:.2e}",  1565), "1.57e3" );
+    assert_eq!(format!("{:.2e}",  1585), "1.59e3" );
+    assert_eq!(format!("{:.2e}",  1605), "1.61e3" );
+    assert_eq!(format!("{:.2e}",  1625), "1.63e3" );
+    assert_eq!(format!("{:.2e}",  1645), "1.65e3" );
+    assert_eq!(format!("{:.2e}",  1665), "1.67e3" );
+    assert_eq!(format!("{:.2e}",  1685), "1.69e3" );
+    assert_eq!(format!("{:.2e}",  1705), "1.71e3" );
+    assert_eq!(format!("{:.2e}",  1725), "1.73e3" );
+    assert_eq!(format!("{:.2e}",  1745), "1.75e3" );
+    assert_eq!(format!("{:.2e}",  1765), "1.77e3" );
+    assert_eq!(format!("{:.2e}",  1785), "1.79e3" );
+    assert_eq!(format!("{:.2e}",  1805), "1.81e3" );
+    assert_eq!(format!("{:.2e}",  1825), "1.83e3" );
+    assert_eq!(format!("{:.2e}",  1845), "1.85e3" );
+    assert_eq!(format!("{:.2e}",  1865), "1.87e3" );
+    assert_eq!(format!("{:.2e}",  1885), "1.89e3" );
+    assert_eq!(format!("{:.2e}",  1905), "1.91e3" );
+    assert_eq!(format!("{:.2e}",  1925), "1.93e3" );
+    assert_eq!(format!("{:.2e}",  1945), "1.95e3" );
+    assert_eq!(format!("{:.2e}",  1965), "1.97e3" );
+    assert_eq!(format!("{:.2e}",  1985), "1.99e3" );
```

Comment:
Aren't we supposed to be rounding ties toward even? So this would be "1.98e3", and many others would round down too.

Reply:
Is round to even spec'd in a doc?  I was trying to keep the same behavior as before using  https://doc.rust-lang.org/std/fmt/#precision as a reference for desired behavior based on the discussion here: https://users.rust-lang.org/t/scientific-notation-decimal-places/99682/5

Reply:
Sort of -- it's specified for operations on the [`f32`](https://doc.rust-lang.org/std/primitive.f32.html) page, and the `f64` page just references that.

> When the number resulting from a primitive operation (addition, subtraction, multiplication, or division) on this type is not exactly representable as f32, it is rounded according to the roundTiesToEven direction defined in IEEE 754-2008.

We do seem to do this for fractional formatting as well:

```rust
fn main() {
    println!("{:.2e}", 1.625);
    println!("{:.2e}", 16.25);
    println!("{:.2e}", 162.5);
    println!("{:.2e}", 1625);
    println!("{:.2e}", 16250);
}
```

Note that my choice of input values are represented precisely at all levels, so there's no ambiguity that these should be ties. The first three do round down to even, but the last two are rounding up.

```
1.62e0
1.62e1
1.62e2
1.63e3
1.63e4
```

Reply:
Ah, I see what you're saying.  Since the bug submitted was the number of digits expressed, would it be ok for me to take out the `assert!`s that enforce the non round-to-even behavior and then work on the rounding in a follow up PR?

Reply:
If you're *willing* to work on both, I'd rather do it all at once so they can land in the same release, and we only have to explain changes once. :)

Reply:
gotcha 🙂 i think what I just pushed rounds to evens on a tie correctly now